sdk/samples/iot/sample_pnp_thermostat_component.c

```diff
@@ -0,0 +1,369 @@
+  AZ_RETURN_IF_FAILED(az_json_writer_append_string(json_builder, start_time_span));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_property_name(json_builder, report_end_time_name_span));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_string(json_builder, end_time_span));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_end_object(json_builder));
+
+  *response_payload = az_json_writer_get_bytes_used_in_destination(json_builder);
+
+  return AZ_OK;
+}
+
+// Invoke the command requested from the service. Here, it generates a report for max, min, and avg
+// temperatures.
+static az_result invoke_getMaxMinReport(
+    sample_pnp_thermostat_component* thermostat_component,
+    az_span payload,
+    az_span response,
+    az_span* out_response)
+{
+  // az_result result;
+  // Parse the "since" field in the payload.
+  az_span start_time_span = AZ_SPAN_NULL;
+  az_json_reader jp;
+  AZ_RETURN_IF_FAILED(az_json_reader_init(&jp, payload, NULL));
+  AZ_RETURN_IF_FAILED(az_json_reader_next_token(&jp));
+  int32_t incoming_since_value_buffer_len;
+  AZ_RETURN_IF_FAILED(az_json_token_get_string(
+      &jp.token,
+      incoming_since_value_buffer,
+      sizeof(incoming_since_value_buffer),
+      &incoming_since_value_buffer_len));
+  start_time_span
+      = az_span_init((uint8_t*)incoming_since_value_buffer, incoming_since_value_buffer_len);
+
+  // Set the response payload to error if the "since" field was not sent
+  if (az_span_ptr(start_time_span) == NULL)
+  {
+    response = report_error_payload;
+    return AZ_ERROR_ITEM_NOT_FOUND;
+  }
+
+  // Get the current time as a string
+  time_t rawtime;
+  struct tm* timeinfo;
+  time(&rawtime);
+  timeinfo = localtime(&rawtime);
+  size_t len = strftime(end_time_buffer, sizeof(end_time_buffer), iso_spec_time_format, timeinfo);
+  az_span end_time_span = az_span_init((uint8_t*)end_time_buffer, (int32_t)len);
+
+  az_json_writer json_builder;
+  AZ_RETURN_IF_FAILED(az_json_writer_init(&json_builder, response, NULL));
+  AZ_RETURN_IF_FAILED(build_command_response_payload(
+      thermostat_component, &json_builder, start_time_span, end_time_span, out_response));
+
+  return AZ_OK;
+}
+
+static az_result build_telemetry_message(
+    sample_pnp_thermostat_component* thermostat_component,
+    az_span payload,
+    az_span* out_payload)
+{
+  az_json_writer json_builder;
+  AZ_RETURN_IF_FAILED(az_json_writer_init(&json_builder, payload, NULL));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_begin_object(&json_builder));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_property_name(&json_builder, telemetry_name));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_double(
+      &json_builder, thermostat_component->current_temperature, DOUBLE_DECIMAL_PLACE_DIGITS));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_end_object(&json_builder));
+  *out_payload = az_json_writer_get_bytes_used_in_destination(&json_builder);
+
+  return AZ_OK;
+}
+
+static az_result append_double(az_json_writer* json_writer, void* value)
```

Comment:
Where are we getting doubles as `void*` from?

Reply:
Callback from the user. The assumption here is if someone passes the `append_double` callback as their callback of choice, then their callback context will be a double pointer.

Reply:
Then, what's stopping us from changing this to be a follows?
```suggestion
static az_result append_double(az_json_writer* json_writer, double* value)
```

Reply:
because the callback is typedef'd to have an `az_json_writer*` and a `void*` as parameters. That way anything can be appended by the user. Say they want to append some complex object. They could do that as long as they passed something like `append_object` and their object as the void pointer. The callback would then know how to handle that.

TLDR: it makes the callback generic enough that the user can append anything instead of just a double.